Maple2.File.Parser/AiParser.cs

```diff
@@ -10,24 +10,265 @@
+                    }
+
+                    field.SetValue(outObject, new Vector3(floatValues[0], floatValues[1], floatValues[2]));
+                    break;
+                case "String":
+                    field.SetValue(outObject, attrib.Value);
+                    break;
+                case "Int32[]":
+                    int[] intValues = Array.ConvertAll(attrib.Value.Split(",", StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries), int.Parse);
+
+                    field.SetValue(outObject, intValues);
+                    break;
+                default:
+                    if (field.FieldType.IsArray && field.FieldType.GetElementType().IsEnum) {
+                        Type? nestedType = field.FieldType.GetElementType();
+                        if (nestedType is null) {
+                            throw new InvalidDataException($"unhandled type of {field.FieldType.Name} on value '{attrib.Value}' for {typeof(T).Name} member {field.Name}");
+                        }
+                        // generates an object[] array of the enums. i couldn't find a way to generate the actual underlying enum type array directly because it kept resolving to object[]
+                        var enumObjectValues = Array.ConvertAll(attrib.Value.Split(",", StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries), value => System.Enum.Parse(nestedType, value));
+                        // forcibly create an enum array and copy values over so that field.SetValue is happy & copies the array over to the member
+                        dynamic enumValues = Array.CreateInstance(field.FieldType.GetElementType(), enumObjectValues.GetLength(0));
+                        for (int i = 0; i < enumObjectValues.Length; ++i) {
+                            enumValues[i] = (dynamic)enumObjectValues[i];
+                        }
+                        field.SetValue(outObject, enumValues);
+                    }
+                    else if (field.FieldType.IsEnum) {
+                        field.SetValue(outObject, System.Enum.Parse(field.FieldType, attrib.Value));
+                    }
+                    else {
+                        throw new InvalidDataException($"unhandled type of {field.FieldType.Name} on value '{attrib.Value}' for {typeof(T).Name} member {field.Name}");
+                    }
+                    break;
+            }
+        }
     }
 
     public IEnumerable<(string aiName, NpcAi Data)> Parse() {
         foreach (PackFileEntry entry in xmlReader.Files.Where(entry => entry.Name.StartsWith("AI/"))) {
             string sanitized = Sanitizer.SanitizeAi(xmlReader.GetString(entry));
-            var root = aiSerializer.Deserialize(XmlReader.Create(new StringReader(sanitized))) as NpcAi;
 
-            Debug.Assert(root != null);
+            XmlDocument document = new XmlDocument();
+            var nodeStack = new List<(int StackIndex, XmlNodeList Nodes, object OutNode)>();
```

Comment:
There is an actual stack class: https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.stack-1?view=net-8.0